examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+	// some thing like that
+
+	// Todo(achal): Look at this:
+	// https://github.com/Devsh-Graphics-Programming/Nabla/blob/6bd5061abe0a2020142efda827269ea6c07f0f2f/examples_tests/common/CommonAPI.h
+	core::smart_refctd_ptr<video::CVulkanPhysicalDevice> gpu = nullptr;
+
+	uint32_t graphicsFamilyIndex(~0u);
+	uint32_t presentFamilyIndex(~0u);
+
+	// Todo(achal): Probably want to put these into some struct
+	nbl::video::ISurface::SFormat surfaceFormat;
+	nbl::video::ISurface::E_PRESENT_MODE presentMode;
+	nbl::video::ISurface::E_SURFACE_TRANSFORM_FLAGS preTransform;
+	nbl::asset::E_SHARING_MODE imageSharingMode;
+	using QueueFamilyIndicesArrayType = core::smart_refctd_dynamic_array<uint32_t>;
+	QueueFamilyIndicesArrayType queueFamilyIndices;
+
+	for (size_t i = 0ull; i < gpus.size(); ++i)
+	{
+		// Todo(achal): Hacks, get rid
+		gpu = core::smart_refctd_ptr_static_cast<video::CVulkanPhysicalDevice>(*(gpus.begin() + i));
+		auto vk_surface = core::smart_refctd_ptr_static_cast<video::ISurfaceVK>(surface);
+
+		bool isGPUSuitable = false;
+
+		// Check if the physical device has queue families which support both graphics and presenting, not necessarily overlapping
+		{
+			const auto& queueFamilyProperties = gpu->getQueueFamilyProperties();
+
+			for (uint32_t familyIndex = 0u; familyIndex < queueFamilyProperties.size(); ++familyIndex)
+			{
+				const auto& familyProperty = queueFamilyProperties.begin() + familyIndex;
+				if (familyProperty->queueFlags & video::IPhysicalDevice::E_QUEUE_FLAGS::EQF_GRAPHICS_BIT)
+					graphicsFamilyIndex = familyIndex;
+
+				if (surface->isSupported(gpu.get(), familyIndex))
+					presentFamilyIndex = familyIndex;
+
+				if ((graphicsFamilyIndex != ~0u) && (presentFamilyIndex != ~0u))
+				{
+					isGPUSuitable = true;
+					break;
+				}
+			}
+		}
+
+		// Check if this physical device supports the swapchain extension
+		// Todo(achal): Eventually move this to somewhere inside the engine
+		{
+			// Todo(achal): Get this from the user
+			const uint32_t requiredDeviceExtensionCount = 1u;
+			const char* requiredDeviceExtensionNames[] = { VK_KHR_SWAPCHAIN_EXTENSION_NAME };
 
+			uint32_t availableExtensionCount;
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, NULL);
+			std::vector<VkExtensionProperties> availableExtensions(availableExtensionCount);
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, availableExtensions.data());
+
+			bool requiredDeviceExtensionsAvailable = false;
+			for (uint32_t i = 0u; i < availableExtensionCount; ++i)
+			{
+				if (strcmp(availableExtensions[i].extensionName, requiredDeviceExtensionNames[0]) == 0)
+				{
+					requiredDeviceExtensionsAvailable = true;
+					break;
+				}
+			}
+
+			if (!requiredDeviceExtensionsAvailable)
+				isGPUSuitable = false;
+		}
+
+		// Check if the surface is adequate
+		{
+			std::vector<video::ISurface::SFormat> surfaceFormats =

```

Comment:
I'd keep to the convention of taking a pointer and a reference to a count, writing out values only if pointer is not null and up to `count`, then overwriting `count` with the actual number of things that could have been written.

Generally speaking its very nasty for a library to return a templated container such as `vector` in its "hard frontend" API because people could be using various different containers and they end up doing nasty reallocations and copies.

Reply:
makes a lot of sense